crates/bevy_ecs/src/schedule/state.rs

```diff
@@ -128,37 +127,166 @@ impl<S: States> NextState<S> {
+            .run_if(common_conditions::run_once())
+            .in_set(StateTransitionSet::RunEnterOnce),
+        (
+            apply_before_transition::<S>.in_set(StateTransitionSet::BeforeTransition),
+            run_on_exit::<S>.in_set(StateTransitionSet::OnExit),
+            run_on_transition::<S>.in_set(StateTransitionSet::OnTransition),
+            run_on_enter::<S>.in_set(StateTransitionSet::OnEnter),
+            apply_after_transition::<S>.in_set(StateTransitionSet::AfterTransition),
+        )
+            .run_if(next_state_changed::<S>),
+    ))
+}
+
+/// Condition used in the beginning of the schedule to check if the next state has changed.
+fn next_state_changed<S: States>(state: Res<State<S>>, next_state: Res<NextState<S>>) -> bool {
+    match &next_state.0 {
+        None => false,
+        Some(next_state) => next_state != state.get(),
+    }
+}
+
+/// Change the state before invoking transition systems.
+fn apply_before_transition<S: States>(
+    mut prev_state: ResMut<PrevState<S>>,
+    mut state: ResMut<State<S>>,
+    mut next_state: ResMut<NextState<S>>,
+) {
+    let entered = next_state.0.take();
+    let entered = entered.expect("NextState<S> should be Some if next_state_changed::<S> is true");
+
+    assert_ne!(*state, entered);
+
+    let exited = mem::replace(&mut state.0, entered);
+    prev_state.0 = Some(exited);
+}
+
+/// If a new state is queued in [`NextState<S>`], this system
+/// runs the [`OnExit(exited_state)`] schedule, if it exists.
+fn run_on_exit<S: States>(world: &mut World) {
+    let exited = world.resource::<PrevState<S>>().0.clone();
+    let exited = exited.expect("PrevState<S> should be Some after apply_before_transition::<S>");
+
+    // Try to run the schedules if they exist.
+    world.try_run_schedule(OnExit(exited)).ok();
+}
+
+/// If a new state is queued in [`NextState<S>`], this system
+/// runs the [`OnTransition { from: exited_state, to: entered_state }`](OnTransition), if it exists.
+fn run_on_transition<S: States>(world: &mut World) {
+    let exited = world.resource::<PrevState<S>>().0.clone();
+    let exited = exited.expect("PrevState<S> should be Some after apply_before_transition::<S>");
+
+    let entered = world.resource::<State<S>>().0.clone();
+
+    // Try to run the schedules if they exist.
+    world
+        .try_run_schedule(OnTransition {
+            from: exited,
+            to: entered,
+        })
+        .ok();
+}
+
+/// If a new state is queued in [`NextState<S>`], this system
+/// runs the [`OnEnter(entered_state)`] schedule, if it exists.
+fn run_on_enter<S: States>(world: &mut World) {
+    let entered = world.resource::<State<S>>().0.clone();
+
+    // Try to run the schedules if they exist.
+    world.try_run_schedule(OnEnter(entered)).ok();
+}
+
+/// Change the state after invoking transition systems.
+fn apply_after_transition<S: States>(mut prev: ResMut<PrevState<S>>) {
+    let prev = prev.0.take();
+
+    assert!(
```

Comment:
I suggested changing to a `debug_assert` since I don't want to cause production builds to crash over something like this. I wonder if we could just make `PrevState` private, then there would be basically no chance of this assertion failing.

Reply:
>  I don't want to cause production builds to crash over something like this

In my opinion it is better to crash at production early than have some other problem in production (including crash in other placer) which is much harder to debug (may be orders of magnitude harder) than assertion (for example, corrupted save files).

> I wonder if we could just make PrevState private, then there would be basically no chance of this assertion failing.

Locally in my project I have `InitResMut`, which is a `ResMut` wrapper which also does `insert_resource` in `init_state`. I think this goes against Bevy model. But it is possible.

Another possibility is mark `PrevState` `#[doc(hidden)]`.

Another possibility is to hide resource initialization in a function like `init_resources_for_state<S: States>(world: &mut World)`.

Reply:
> In my opinion it is better to crash at production early than have some other problem in production (including crash in other placer) which is much harder to debug (may be orders of magnitude harder) than assertion (for example, corrupted save files).

The way I see it, the goal of a production build is not to be debuggable. It's to provide a reliable and smooth user experience. In release builds, it's sometimes better to keep chugging along in an invalid state rather than crashing the whole application, especially when the invalid state is as minor as this one.